File: summary/rust/2_merge_sort.rs
fn merge_sort<T: PartialOrd + Copy>(arr: &mut [T]) {
    let mid = arr.len() / 2;
    if mid == 0 {
        return;
    }

    merge_sort(&mut arr[..mid]);
    merge_sort(&mut arr[mid..]);

    let mut tmp = vec![arr[0]; arr.len()];
    merge(&arr[..mid], &arr[mid..], &mut tmp[..]);
    arr.copy_from_slice(&tmp);
}

fn merge<T: PartialOrd + Copy>(left: &[T], right: &[T], result: &mut [T]) {
    let mut i = 0;
    let mut j = 0;
    let mut k = 0;

    while i < left.len() && j < right.len() {
        if left[i] < right[j] {
            result[k] = left[i];
            i += 1;
        } else {
            result[k] = right[j];
            j += 1;
        }
        k += 1;
    }

    if i < left.len() {
        result[k..].copy_from_slice(&left[i..]);
    } else {
        result[k..].copy_from_slice(&right[j..]);
    }
}

fn main() {
    let mut arr = [5, 3, 8, 1, 7];
    merge_sort(&mut arr);
    println!("{:?}", arr); // [1, 3, 5, 7, 8]
}

File: summary/dart/2_merge_sort.dart
List<int> mergeSort(List<int> arr) {
  if (arr.length <= 1) {
    return arr;
  }

  int mid = arr.length ~/ 2;
  List<int> left = mergeSort(arr.sublist(0, mid));
  List<int> right = mergeSort(arr.sublist(mid));

  return merge(left, right);
}

List<int> merge(List<int> left, List<int> right) {
  List<int> result = [];
  int i = 0, j = 0;

  while (i < left.length && j < right.length) {
    if (left[i] <= right[j]) {
      result.add(left[i++]);
    } else {
      result.add(right[j++]);
    }
  }

  while (i < left.length) {
    result.add(left[i++]);
  }

  while (j < right.length) {
    result.add(right[j++]);
  }

  return result;
}

void main() {
  List<int> numbers = [34, 19, 42, -9, 2018, 0, 2005, 77, 209];
  List<int> sortedNumbers = mergeSort(numbers);
  print('Sorted list: $sortedNumbers');
}
